### PR TITLE
🔊 Add log to warn when SDK user agent is missing

### LINKED ISSUE #455: Log warning if no env or client info are passed
We use those two strings for debugging SDKs. I think we should log a warning if at any point that info is missing from an implemented SDK. Without the user agent info, it's pretty hard to know which SDK / test framework the snapshots are coming from.

### PR DESCRIPTION
**Pull request by Robdel12**
## What is this?

This implements a warning log if the SDK sends a snapshot without any user agent info (`environmentInfo` & `clientInfo`). Closes #455 

Before releasing this, we will need to update the Storybook SDK

### CODE REVIEW

File: packages/cli-build/src/commands/build/wait.js [hunk 1]
```diff
@@ -56,10 +57,12 @@ export class Wait extends Command {
       return this.log.info('Percy is disabled');
     }
 
-    await new PercyClient()
-      .waitForBuild(this.flags, data => {
-        this.status(data);
-      });
+    await new PercyClient({
+      clientInfo: `${pkg.name}/${pkg.version}`,
+      environmentInfo: ''
```
Comment (Robdel12) [inline packages/cli-build/src/commands/build/wait.js:62]: @wwilsman 🚌 Strugglin' to think of what should go here. The node version is passed by the client always, so I'm not sure what would be useful here.

Also feels not great an empty string is all it takes to get around this lol

Comment (wwilsman) [reply packages/cli-build/src/commands/build/wait.js:62]: You're right, empty environment info doesn't feel correct, but also in these instances it's already handled since client adds Node env info automatically.

I was going to suggest adding node info anyway since a `Set` is used and values can only ever be unique. However, I realized that when constructing the user-agent, node info is always added regardless if it already exists (which feels like a bug). This comment is starting to stray from this line of code, so I'll move it to the proper place. 😆

File: packages/client/src/client.js [hunk 2]
```diff
@@ -31,13 +31,14 @@ function validateProjectPath(path) {
...
+  envWarning = false;
 
   constructor({
     // read or write token, defaults to PERCY_TOKEN environment variable
     token,
     // initial user agent info
-    clientInfo = '',
-    environmentInfo = '',
+    clientInfo = [],
+    environmentInfo = [],
```
Comment (wwilsman) [inline packages/client/src/client.js:41]: (I can't comment on the unchanged folded code below this, but I'll comment here because it's close enough)

Rather than initialize these properties in this constructor, we can move the `new Set()`s to the same place you added the `envWarning`. Then in the constructor, utilize the `this.add*Info()` methods to handle empty strings (which they already do).

File: packages/client/src/client.js [hunk 3]
```diff
@@ -67,6 +68,11 @@ export default class PercyClient {
 
   // Stringifies client and environment info.
   userAgent() {
+    if (!this.envWarning && (this.clientInfo.size === 0 || this.environmentInfo.size === 0)) {
```
Comment (wwilsman) [inline packages/client/src/client.js:71]: By moving the initial values and utilizing the existing methods, at this point we should be certain that these Sets only contain truthy and unique values. This way empty strings will no longer work around the warning.

File: packages/client/src/client.js [hunk 4]
```diff
@@ -67,6 +68,11 @@ export default class PercyClient {
...
+      this.envWarning = true;
+      this.log.warn('Warning: Missing `clientInfo` and/or `environmentInfo` properties');
+    }
+
     let client = [`Percy/${/\w+$/.exec(this.apiUrl)}`]
       .concat(`${pkg.name}/${pkg.version}`, ...this.clientInfo)
       .filter(Boolean).join(' ');
```
Comment (wwilsman) [inline packages/client/src/client.js:78]: The forced client info here is safe to assume will always be unique since no other SDK should be providing API or `@percy/client` info. However, if we wanted to be sure it is unique, we'd have to do the same as my suggestion below, only because of the order of forced client info, we would also need to reverse the resulting array.

(Again, I can't comment on unchanged folded code below)

For environment info below, we can add the forced info to the Set so it's always unique. I would suggest cloning the set to avoid mutation, but considering the whole point of Sets is to have unique values, it's safe to just keep adding the forced info blindly. We still want the variable to be an array though, so we can just move the `concat` into `[...envInfo.add(...).add(...)]`. (I really wish GH would let me comment on folded code)

File: packages/cli-build/src/commands/build/finalize.js [hunk 5]
```diff
@@ -26,7 +27,10 @@ export class Finalize extends Command {
       process.env.PERCY_PARALLEL_TOTAL = '-1';
     }
 
-    let client = new PercyClient();
+    let client = new PercyClient({
+      clientInfo: `${pkg.name}/${pkg.version}`,
+      environmentInfo: ''
```
Comment (wwilsman) [inline packages/cli-build/src/commands/build/finalize.js:32]: Now back to these. With those changes, we should be safe to repeat node info. It might help rule out any crazy environments if multiple node versions show up.

**Review by wwilsman — CHANGES_REQUESTED** [review]
I agree with your empty string workaround, but my suggested fix revealed a bug. So read on to see my suggestion to fix it. 😆

File: packages/client/src/client.js (hunk 4 above)
Comment (wwilsman) [reply packages/client/src/client.js:78]: Also, if we forcibly add both client and env info, the `envWarning` won't be needed since we know an empty value would only ever happen the first time

File: packages/client/test/client.test.js [hunk 6]
```diff
@@ -3,12 +3,15 @@ import { mockgit } from '@percy/env/test/helpers';
...
 import { sha256hash, base64encode } from '../src/utils';
 import mockAPI from './helpers';
+import logger from '@percy/logger/test/helpers';
 
 describe('PercyClient', () => {
   let client;
 
   beforeEach(() => {
     mockAPI.start();
+    logger.mock();
```
Comment (Robdel12) [inline packages/client/test/client.test.js:13]: ✂️  dead

File: packages/client/test/client.test.js [hunk 7]
```diff
@@ -3,12 +3,15 @@ import { mockgit } from '@percy/env/test/helpers';
 import PercyClient from '../src';
 import { sha256hash, base64encode } from '../src/utils';
 import mockAPI from './helpers';
+import logger from '@percy/logger/test/helpers';
```
Comment (Robdel12) [inline packages/client/test/client.test.js:6]: v

File: packages/core/src/percy.js [hunk 8]
```diff
@@ -235,6 +236,11 @@ export default class Percy {
       throw new Error('Not running');
     }
 
+    if (!this.envWarning && (!options.clientInfo || !options.environmentInfo)) {
+      this.envWarning = true;
+      this.log.warn('Warning: Missing `clientInfo` and/or `environmentInfo` properties');
+    }
+
```
Comment (wwilsman) [inline packages/core/src/percy.js:243]: Thinking about this and seeing other changes, it doesn't quite feel like the right place to warn.

For example, some SDKs iterate over a list and use the `snapshot` method directly. They correctly provide env and client info, but they don't provide it with snapshots. So this warning would incorrectly trigger. 

As another example, the tests that you've changed previously had valid use cases where you can use the snapshot method without client/env info. If that info were to be required for every snapshot, this whole PR would be made invalid by adding the properties to the validation schema as required properties.

Now that I'm noticing this warning, I believe my previous comment about the env info being in the wrong place in the snapshot command was due to this warning. In that command, the env/client info should be passed on init and not with every snapshot. This also explains why you needed to duplicate the client info that was already present.

A better place for this would be in client. Although when discussing this offline, we discovered it was too eager to do so at the construction of the UA header because we didn't want to trigger the warning when a build is created. I think we're on the right path moving it to the same place snapshots are handled, but instead, we should move it to the `client.createSnapshot` method. That way we avoid the check for builds, but still check at the last moment before it should be there.

**Review by wwilsman — CHANGES_REQUESTED** [review]

File: packages/client/src/client.js [hunk 9]
```diff
@@ -31,19 +31,27 @@ function validateProjectPath(path) {
...
-    this.clientInfo = new Set([].concat(clientInfo));
-    this.environmentInfo = new Set([].concat(environmentInfo));
+    // set our env info
+    this.addClientInfo(`Percy/${/\w+$/.exec(this.apiUrl)}`);
+    this.addClientInfo(`${pkg.name}/${pkg.version}`);
+    this.addEnvironmentInfo(`node/${process.version}`);
+    this.addEnvironmentInfo(this.env.info);
+    // set info passed from the SDK
+    this.addClientInfo([].concat(clientInfo).filter(Boolean));
+    this.addEnvironmentInfo([].concat(environmentInfo).filter(Boolean));
```
[the remark this answers is not in the document]
Comment (wwilsman) [reply packages/client/src/client.js:54]: With my other review comment below, that check wouldn't work since we're adding our own info as soon as possible here. We'll have to go back to adding this all lazily when the UA is created. To avoid duplicates (which was the goal of moving these here), we can duplicate the sets immutably with our information appended first.

File: packages/cli-snapshot/src/commands/snapshot.js [hunk 10]
```diff
@@ -123,7 +124,7 @@ export class Snapshot extends Command {
           this.log.debug(`-> url: ${snap.url}`);
         }
       } else {
-        this.percy.snapshot(snap);
+        this.percy.snapshot({ ...snap });
```
Comment (wwilsman) [inline packages/cli-snapshot/src/commands/snapshot.js:127]: ✂️  Can revert this change

File: packages/client/src/client.js [hunk 11]
```diff
@@ -292,6 +296,11 @@ export default class PercyClient {
     this.addClientInfo(clientInfo);
     this.addEnvironmentInfo(environmentInfo);
 
+    if (!this.envWarning && (this.clientInfo.size === 0 || this.environmentInfo.size === 0)) {
+      this.envWarning = true;
+      this.log.warn('Warning: Missing `clientInfo` and/or `environmentInfo` properties');
+    }
```
Comment (wwilsman) [inline packages/client/src/client.js:302]: Small nit
```suggestion
    if (!this.envWarning && (!this.clientInfo.size || !this.environmentInfo.size)) {
      this.log.warn('Warning: Missing `clientInfo` and/or `environmentInfo` properties');
      this.envWarning = true;
    }
```

AND/OR, how annoying do we want this warning / what do we think of logging this for every snapshot (getting rid of the `envWarning` variable)? I'm wondering how easy this might be to miss if only logged once (plus this shouldn't be an issue for our own SDKs (and if so, we need to fix them anyway))

File: packages/core/test/percy.test.js [hunk 12]
```diff
@@ -356,7 +362,9 @@ describe('Percy', () => {
       await percy.snapshot({
         name: 'test snapshot',
         url: 'http://localhost:8000',
-        domSnapshot: '<html></html>'
+        domSnapshot: '<html></html>',
+        clientInfo: 'client-info',
+        environmentInfo: 'env-info'
       });
 
       await expectAsync(percy.stop()).toBeResolved();
```
Comment (wwilsman) [inline packages/core/test/percy.test.js:370]: ✂️  Can revert this file

File: packages/core/test/snapshot.test.js [hunk 13]
```diff
@@ -546,7 +581,9 @@ describe('Snapshot', () => {
         execute: {
           beforeResize: domtest('beforeResize', () => window.innerWidth),
           afterResize: domtest('afterResize', () => window.innerWidth)
-        }
+        },
+        clientInfo: 'client-info',
+        environmentInfo: 'env-info'
       });
 
       await percy.idle();
```
Comment (wwilsman) [inline packages/core/test/snapshot.test.js:589]: ✂️  This file too

**Review by wwilsman — CHANGES_REQUESTED** [review]

File: packages/core/test/percy.test.js (hunk 12 above)
Comment (wwilsman) [reply packages/core/test/percy.test.js:370]: Oh, maybe we need to add these to the `new Percy` init in the before each?

File: packages/core/test/snapshot.test.js (hunk 13 above)
Comment (wwilsman) [reply packages/core/test/snapshot.test.js:589]: Same comment here about moving to `new Percy` 😆

**Review by wwilsman — APPROVED** [review]
🎮
